### DISCUSSION TITLE
Allow awaiting nullable tasks

### DISCUSSION THREAD

**Discussion by MrTA2** [dotnet/csharplang#4007] [Language Ideas] [answered]
Consider this code snipper:

```
Func<Task>? del = GetDelegate();
await del?.Invoke();

IAsyncDisposable? obj = GetDisposable();
await obj?.DisposeAsync();
```

First version generates a warning and an NRE at runtime, and the second one generates an error, because the method call expression, not having been executed, returns `Task? null`. 

Proposed solution: await adds a check for Task? and ValueTask? - await if not null, do nothing if null.

**Comment by HaloFour** [comment, marked as answer]
Check out: https://github.com/dotnet/csharplang/issues/35

**Reply by MrTA2** [reply]
Thanks!
